### ISSUE TITLE
A client that falls back to a different endpoint on error

### ISSUE THREAD

**Issue by bajtos** [drand/drand-client#84]
In Spark, we were recently affected by a temporary outage of `https://api.drand.sh/` (see https://github.com/filecoin-station/spark/issues/90). Other endpoints listed in [DevGuide: HTTP API](https://drand.love/docs/http-api-reference) were fully operational.

It would be great if `drand-client` could provide a chain/client implementation that accepts a list of endpoint URLs and falls back to the second/third/... provider in the list when the requested endpoint returns an 5xx error.

Example usage:

```js
const options = {
  // ...
  chainVerificationParams: { chainHash, publicKey }
}

const urls = [
    'https://api.drand.sh',
    'https://drand.cloudflare.com'
    // ...
]
const client = new FallbackClient(urls, options)
const theLatestBeacon = await fetchBeacon(client)
```

**Comment by CluEleSsUK** [comment]
Hi @bajtos, this already exists [here](https://github.com/drand/drand-client/blob/master/lib/fastest-node-client.ts). Nodes that go down will get an infinite response time and go to the back of the list until they’re back up.

Feel free to reopen or create a new issue if this isn’t sufficient for your needs

**Comment by bajtos** [comment]
> Hi @bajtos, this already exists [here](https://github.com/drand/drand-client/blob/master/lib/fastest-node-client.ts). Nodes that go down will get an infinite response time and go to the back of the list until they’re back up.
> 
> Feel free to reopen or create a new issue if this isn’t sufficient for your needs

I was aware of `FastestNodeClient`, but it seems to do something else: it runs the test speed only once, when you call `start`. To get the desired behaviour, I would have to call `start` every time I want to query drand. That would send a lot of unnecessary requests to DRAND API providers. Plus, there is no guarantee that the fastest endpoint to respond to the speed test will be still up & running when the next request arrives.

Specifically, I want a solution to handle situations like the short outage of https://api.drand.sh/ that happened a few weeks ago.

We can implement this on our side, but I thought such a feature could also be useful for other DRAND API consumers.

**Comment by kentbull** [comment]
So essentially a client that fails over to the next available node, possibly in a round robin order?

**Comment by CluEleSsUK** [comment]
The `FastestNodeClient` shouldn’t run only once, but on every interval provided as per the `setInterval` here: https://github.com/drand/drand-client/blob/d97088d08dabe5f04b11d21b9600d4becb75c324/lib/speedtest.ts#L28
you ought to only need to call `start` once unless something is very broken

**Comment by bajtos** [comment]
> So essentially a client that fails over to the next available node, possibly in a round robin order?

Yes, exactly!

> The `FastestNodeClient` shouldn’t run only once, but on every interval provided as per the `setInterval` here:
> 
> https://github.com/drand/drand-client/blob/d97088d08dabe5f04b11d21b9600d4becb75c324/lib/speedtest.ts#L28
> 
> you ought to only need to call `start` once unless something is very broken

Ah, I missed that part.

To handle short outages, I would have to configure FastestNodeClient with a short speed test interval, e.g., one second, and implement retry logic for failed requests with a delay between requests larger than the speed test interval. That sounds like a lot of complexity! Plus, the frequent speed checks create a lot of load on the backup providers.

**Comment by kentbull** [comment]
I will make sure this is added to the docs. I am sure other people will have the same question.
